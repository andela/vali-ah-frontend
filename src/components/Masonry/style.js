import styled from 'styled-components';

export const MasonryDiv = styled.div`
  display: grid;
  grid-auto-flow: column;
  grid-gap: ${(props) => props.gap || '.1em'};
`;

export const Col = styled.div`
  display: grid;
  grid-gap: ${(props) => props.gap || '.1em'};
  grid-auto-rows: unset;
`;

export const Margin = styled.div`
  margin-top: ${(props) => `${props.margin}px`};
`;
