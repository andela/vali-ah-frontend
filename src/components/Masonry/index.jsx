import React, { useRef, useState, useEffect } from 'react';
import Proptypes from 'prop-types';

import { MasonryDiv, Col, Margin } from './style';

const Masonry = ({ children, gap, minWidth = 300 }) => {
  const cols = [];
  const ref = useRef();
  const [numCols, setNumCols] = useState(3);

  const calcNumCols = () => setNumCols(Math.floor(ref.current.offsetWidth / minWidth));

  const createCols = () => {
    if (!ref.current) return;
    const randIndex = () => Math.floor(Math.random() * 30);
    // eslint-disable-next-line no-plusplus
    for (let i = 0; i < numCols; i++) cols[i] = [];
    children
      .forEach((Child, i) => cols[i % numCols].push(
        <Margin margin={randIndex(i)}>{Child}</Margin>
      ));
  };


  useEffect(() => {
    window.addEventListener('resize', calcNumCols);

    return () => window.removeEventListener('resize', calcNumCols);
  });

  createCols();

  return (
    <MasonryDiv ref={ref} gap={gap}>
      {Array(numCols)
        .fill()
        .map((el, index) => (
          // eslint-disable-next-line react/no-array-index-key
          <Col key={index} gap={gap}>
            {cols[index]}
          </Col>
        ))}
    </MasonryDiv>
  );
};

Masonry.propTypes = {
  children: Proptypes.node,
  gap: Proptypes.number,
  minWidth: Proptypes.number
};

Masonry.defaultProps = {
  children: '',
  gap: '',
  minWidth: 300
};

export default Masonry;
