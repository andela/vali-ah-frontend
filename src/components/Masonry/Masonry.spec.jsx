import React from 'react';

import Masonry from './index';


describe('Masonry Component', () => {
  it('should renders properly without card data', () => {
    const masonrySection = mount(
      <Masonry>
        <span>Span 1</span>
        <span>Span 1</span>
        <span>Span 1</span>
      </Masonry>
    );

    expect(masonrySection).toMatchSnapshot();
  });
});
