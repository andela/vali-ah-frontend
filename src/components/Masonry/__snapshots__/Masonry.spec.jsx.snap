// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masonry Component should renders properly without card data 1`] = `
<Masonry
  gap=""
  minWidth={300}
>
  <styled.div
    gap=""
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "keyKcu",
            "rules": Array [
              "
  display: grid;
  grid-auto-flow: column;
  grid-gap: ",
              [Function],
              ";
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={
        Object {
          "current": <div
            class="sc-bdVaJa keyKcu"
          >
            <div
              class="sc-bwzfXH itrbwG"
            />
            <div
              class="sc-bwzfXH itrbwG"
            />
            <div
              class="sc-bwzfXH itrbwG"
            />
          </div>,
        }
      }
      gap=""
    >
      <div
        className="sc-bdVaJa keyKcu"
      >
        <styled.div
          gap=""
          key="0"
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "itrbwG",
                  "rules": Array [
                    "
  display: grid;
  grid-gap: ",
                    [Function],
                    ";
  grid-auto-rows: unset;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            gap=""
          >
            <div
              className="sc-bwzfXH itrbwG"
            />
          </StyledComponent>
        </styled.div>
        <styled.div
          gap=""
          key="1"
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "itrbwG",
                  "rules": Array [
                    "
  display: grid;
  grid-gap: ",
                    [Function],
                    ";
  grid-auto-rows: unset;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            gap=""
          >
            <div
              className="sc-bwzfXH itrbwG"
            />
          </StyledComponent>
        </styled.div>
        <styled.div
          gap=""
          key="2"
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "itrbwG",
                  "rules": Array [
                    "
  display: grid;
  grid-gap: ",
                    [Function],
                    ";
  grid-auto-rows: unset;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            gap=""
          >
            <div
              className="sc-bwzfXH itrbwG"
            />
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</Masonry>
`;
